pkg/virt-handler/vm.go

```diff
@@ -1270,16 +1304,31 @@ func (d *VirtualMachineController) handleMigrationProxy(vmi *v1.VirtualMachineIn
 
 	// handle starting/stopping source migration proxy.
 	// start the source proxy once we know the target address
-	if d.isMigrationSource(vmi) {
-		err := d.migrationProxy.StartSourceListener(string(vmi.UID), vmi.Status.MigrationState.TargetNodeAddress)
-		if err != nil {
-			return err
+	ports := map[int]int{0: 0}
+	if vmi.Status.MigrationState != nil {
+		if vmi.Status.MigrationState.TargetDirectMigrationNodePorts == nil {
+			log.Log.Object(vmi).Warning("No migration proxy has been created for this vmi")
+			return nil
 		}
+		ports = vmi.Status.MigrationState.TargetDirectMigrationNodePorts
 
-	} else {
-		d.migrationProxy.StopSourceListener(string(vmi.UID))
 	}
 
+	for srcPort, destPort := range ports {
+		key := string(vmi.UID)
+		if srcPort != 0 {
+			key += fmt.Sprintf("-%d", srcPort)
+		}
+		if d.isMigrationSource(vmi) {
+			targetAddr := fmt.Sprintf("%s:%d", vmi.Status.MigrationState.TargetNodeAddress, destPort)
+			err := d.migrationProxy.StartSourceListener(key, targetAddr)
```

Comment:
I had a thought browsing through the code. It might be something you already have planned.

The migrationProxy interface was originally designed with the assumption that there's a single port for each source and target proxy. The "key" was meant to represent all the connections associate with the VMI. 

Now that each VMI can have multiple src/target proxies opened during a migration, we might want consider enhancing the migrationProxy abstraction so each key can manage multiple ports. This would be helpful in the processVmCleanup() function where it's possible we no longer have any state to provide hints regarding what migration connections were created.

Reply:
Oh, absolutely. I need to find a way to improve this. So far I was just focusing on just making everything work.

Thanks!